File: Perez-Meneses-Juan-Carlos/tasks_opt/fc.c
#include <pthread.h>
#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define NUM_FILOSOFOS 5

// Semáforos para los tenedores y el mutex
sem_t tenedores[NUM_FILOSOFOS];
sem_t mutex; // Controla el acceso para evitar deadlocks

void *filosofo(void *num) {
    int id = *(int *)num;

    while (1) {
        printf("Filósofo %d está pensando 🤔\n", id);
        sleep(rand() % 3 + 1); // Tiempo para pensar

        // Intentar tomar tenedores
        sem_wait(&mutex); // Controlar que solo 4 filósofos intenten comer a la vez
        sem_wait(&tenedores[id]); // Tomar el tenedor izquierdo
        sem_wait(&tenedores[(id + 1) % NUM_FILOSOFOS]); // Tomar el tenedor derecho

        printf("Filósofo %d está comiendo 🍝\n", id);
        sleep(rand() % 3 + 1); // Tiempo para comer

        // Soltar tenedores
        sem_post(&tenedores[id]); // Soltar el tenedor izquierdo
        sem_post(&tenedores[(id + 1) % NUM_FILOSOFOS]); // Soltar el tenedor derecho
        sem_post(&mutex); // Permitir que otro filósofo intente comer

        printf("Filósofo %d ha terminado de comer y está pensando nuevamente 🤔\n", id);
    }

    pthread_exit(NULL);
}

int main() {
    pthread_t hilos[NUM_FILOSOFOS];
    int id[NUM_FILOSOFOS];

    // Inicializar semáforos
    sem_init(&mutex, 0, NUM_FILOSOFOS - 1); // Solo 4 filósofos pueden comer simultáneamente
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        sem_init(&tenedores[i], 0, 1); // Inicializar cada tenedor como disponible
    }

    // Crear hilos para cada filósofo
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        id[i] = i;
        pthread_create(&hilos[i], NULL, filosofo, (void *)&id[i]);
    }

    // Esperar a que terminen los hilos (en este caso, nunca terminarán)
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        pthread_join(hilos[i], NULL);
    }

    // Destruir semáforos (nunca se llega aquí en este ejemplo)
    for (int i = 0; i < NUM_FILOSOFOS; i++) {
        sem_destroy(&tenedores[i]);
    }
    sem_destroy(&mutex);

    return 0;
}

File: Perez-Meneses-Juan-Carlos/tasks_opt/fc.py
import threading
import time
import random

# Número de filósofos
NUM_FILOSOFOS = 5
tenedores = [threading.Semaphore(1) for _ in range(NUM_FILOSOFOS)]
mutex = threading.Semaphore(4)  # Solo 4 filósofos pueden intentar comer a la vez

def filosofo(num):
    while True:
        print(f"Filósofo {num} está pensando 🤔")
        time.sleep(random.uniform(1, 3))  # Tiempo de pensar
        
        # Intentar tomar tenedores
        mutex.acquire()
        tenedores[num].acquire()
        tenedores[(num + 1) % NUM_FILOSOFOS].acquire()
        
        print(f"Filósofo {num} está comiendo 🍝")
        time.sleep(random.uniform(1, 3))  # Tiempo de comer
        
        # Soltar tenedores
        tenedores[num].release()
        tenedores[(num + 1) % NUM_FILOSOFOS].release()
        mutex.release()
        
        print(f"Filósofo {num} ha terminado de comer y está pensando nuevamente 🤔")

# Crear hilos para cada filósofo
hilos = [threading.Thread(target=filosofo, args=(i,)) for i in range(NUM_FILOSOFOS)]
for hilo in hilos:
    hilo.start()
for hilo in hilos:
    hilo.join()
    
    
    
"""

Resultados:

ilósofo 0 está pensando 🤔
Filósofo 1 está pensando 🤔
Filósofo 3 está pensando 🤔
Filósofo 2 está pensando 🤔
Filósofo 4 está pensando 🤔
Filósofo 1 está comiendo 🍝
Filósofo 1 ha terminado de comer y está pensando nuevamente 🤔
Filósofo 1 está pensando 🤔
Filósofo 0 está comiendo 🍝
Filósofo 0 ha terminado de comer y está pensando nuevamente 🤔
Filósofo 0 está pensando 🤔
Filósofo 4 está comiendo 🍝
Filósofo 4 ha terminado de comer y está pensando nuevamente 🤔
Filósofo 4 está pensando 🤔
Filósofo 3 está comiendo 🍝
Filósofo 3 ha terminado de comer y está pensando nuevamente 🤔
Filósofo 3 está pensando 🤔
Filósofo 2 está comiendo 🍝

"""
